src/Compilers/Core/Portable/SourceGeneration/IncrementalGeneratorInitializationContext_ForAttributeWithMetadataName.cs

```diff
@@ -0,0 +1,167 @@
+{
+    private static readonly char[] s_nestedTypeNameSeparators = new char[] { '+' };
+    private static readonly SymbolDisplayFormat s_metadataDisplayFormat =
+        SymbolDisplayFormat.QualifiedNameArityFormat.AddCompilerInternalOptions(SymbolDisplayCompilerInternalOptions.UsePlusForNestedTypes);
+
+    /// <summary>
+    /// Creates an <see cref="IncrementalValuesProvider{T}"/> that can provide a transform over all <see
+    /// cref="SyntaxNode"/>s if that node has an attribute on it that binds to a <see cref="INamedTypeSymbol"/> with the
+    /// same fully-qualified metadata as the provided <paramref name="fullyQualifiedMetadataName"/>. <paramref
+    /// name="fullyQualifiedMetadataName"/> should be the fully-qualified, metadata name of the attribute, including the
+    /// <c>Attribute</c> suffix.  For example <c>"System.CLSCompliantAttribute</c> for <see
+    /// cref="System.CLSCompliantAttribute"/>.
+    /// </summary>
+    /// <param name="predicate">A function that determines if the given <see cref="SyntaxNode"/> attribute target (<see
+    /// cref="GeneratorAttributeSyntaxContext.AttributeTarget"/>) should be transformed.  Nodes that do not pass this
+    /// predicate will not have their attributes looked at at all.</param>
+    /// <param name="transform">A function that performs the transform. This will only be passed nodes that return <see
+    /// langword="true"/> for <paramref name="predicate"/> and which have a matchin <see cref="AttributeData"/> whose
+    /// <see cref="AttributeData.AttributeClass"/> has the same fully qualified, metadata name as <paramref
+    /// name="fullyQualifiedMetadataName"/>.</param>
+    internal IncrementalValuesProvider<T> ForAttributeWithMetadataName<T>(
+        string fullyQualifiedMetadataName,
+        Func<SyntaxNode, CancellationToken, bool> predicate,
+        Func<GeneratorAttributeSyntaxContext, CancellationToken, T> transform)
+    {
+        var metadataName = fullyQualifiedMetadataName.Contains('+')
+            ? MetadataTypeName.FromFullName(fullyQualifiedMetadataName.Split(s_nestedTypeNameSeparators).Last())
+            : MetadataTypeName.FromFullName(fullyQualifiedMetadataName);
+
+        var nodesWithAttributesMatchingSimpleName = this.ForAttributeWithSimpleName(metadataName.UnmangledTypeName, predicate);
+
+        var collectedNodes = nodesWithAttributesMatchingSimpleName
+            .Collect()
+            .WithComparer(ImmutableArrayValueComparer<SyntaxNode>.Instance)
+            .WithTrackingName("collectedNodes_ForAttributeWithMetadataName");
+
+        // Group all the nodes by syntax tree, so we can process a whole syntax tree at a time.  This will let us make
+        // the required semantic model for it once, instead of potentially many times (in the rare, but possible case of
+        // a single file with a ton of matching nodes in it).
+        var groupedNodes = collectedNodes.SelectMany(
```

Comment:
I think this is ok for now, as the API shape should be the same, but comes back to what I mentioned about using the SyntaxStrategy, so that all of the syntax inputs of whatever kind share the same semantic model.

Reply:
i'm not sure what this means.